libkbfs/tlf_journal.go

```diff
@@ -851,34 +873,107 @@ func (j *tlfJournal) removeFlushedBlockEntries(ctx context.Context,
 }
 
 func (j *tlfJournal) flushBlockEntries(
-	ctx context.Context, end journalOrdinal) (int, MetadataRevision, error) {
+	ctx context.Context, end journalOrdinal) (
+	int, MetadataRevision, bool, error) {
 	entries, maxMDRevToFlush, err := j.getNextBlockEntriesToFlush(ctx, end)
 	if err != nil {
-		return 0, MetadataRevisionUninitialized, err
+		return 0, MetadataRevisionUninitialized, false, err
 	}
 
 	if entries.length() == 0 {
-		return 0, maxMDRevToFlush, nil
+		return 0, maxMDRevToFlush, false, nil
 	}
 
 	j.log.CDebugf(ctx, "Flushing %d blocks, up to rev %d",
 		len(entries.puts.blockStates), maxMDRevToFlush)
 
+	// Mark these blocks as flushing, and clear when done.
+	err = j.markFlushingBlockIDs(entries)
+	if err != nil {
+		return 0, MetadataRevisionUninitialized, false, err
+	}
+	cleared := false
+	defer func() {
+		if !cleared {
+			_ = j.clearFlushingBlockIDs(entries)
+		}
+	}()
+
 	// TODO: fill this in for logging/error purposes.
 	var tlfName CanonicalTlfName
-	err = flushBlockEntries(ctx, j.log, j.delegateBlockServer,
-		j.config.BlockCache(), j.config.Reporter(),
-		j.tlfID, tlfName, entries)
+
+	eg, groupCtx := errgroup.WithContext(ctx)
+	convertCtx, convertCancel := context.WithCancel(groupCtx)
+
+	// Flush the blocks in a goroutine. Alongside make another
+	// goroutine that listens for MD puts and checks the size of the
+	// MD journal, and converts to a local squash branch if it gets
+	// too large.  While the 2nd goroutine is waiting, it should exit
+	// immediately as soon as the 1st one finishes, but if it is
+	// already working on a conversion it should finish that work.
+	//
+	// If the 2nd goroutine makes a branch, foreground writes could
+	// trigger CR while blocks are still being flushed.  This can't
+	// usually happen, because flushing is paused while CR is
+	// happening.  flush() has to make sure to get the new MD journal
+	// end, and we need to make sure `maxMDRevToFlush` is still valid.
+	eg.Go(func() error {
+		defer convertCancel()
+		return flushBlockEntries(groupCtx, j.log, j.delegateBlockServer,
+			j.config.BlockCache(), j.config.Reporter(),
+			j.tlfID, tlfName, entries)
+	})
+	converted := false
+	eg.Go(func() error {
+		for {
+			select {
+			case <-j.checkMDLenCh:
+				// Don't signal a pause when doing this conversion in
+				// a separate goroutine, because it ends up canceling
+				// the flush context, which means all the block puts
+				// would get canceled and we don't want that.
+				// Instead, let the current iteration of the flush
+				// finish, and then signal at the top of the next
+				// iteration.
+				convertedNow, err :=
+					j.convertMDsToBranchIfOverThreshold(groupCtx, false)
+				if err != nil {
+					return err
+				}
+				converted = converted || convertedNow
```

Comment:
No, because CR can run to completion, and 20 more revisions could be made locally, all while the blocks are flushing.  And over and over again, depending on the speed of the local CPU/disk vs. the speed of the network.